python/mqt/core/qdmi/qiskit/capabilities.py

```diff
@@ -0,0 +1,351 @@
+            is_zone=s.is_zone(),
+            x_extent=s.x_extent(),
+            y_extent=s.y_extent(),
+            z_extent=s.z_extent(),
+            module_index=s.module_index(),
+            submodule_index=s.submodule_index(),
+            t1=s.t1(),
+            t2=s.t2(),
+        )
+    except Exception:  # noqa: BLE001
+        return None
+
+
+def _safe_operation_info(op: fomac.Device.Operation) -> DeviceOperationInfo | None:
+    """Safely build a DeviceOperationInfo for an operation or return None.
+
+    Interprets the sites list according to QDMI specification:
+    - For local multi-qubit operations: returns tuples of site combinations
+    - For single-qubit and global operations: returns individual sites
+
+    Returns:
+        A populated DeviceOperationInfo instance, or ``None`` if extraction failed.
+    """
+    name = op.name()
+    site_list = op.sites()
+
+    # Determine how to interpret the sites based on operation type
+    qubits_num = op.qubits_num()
+    is_zoned = op.is_zoned()
+
+    site_indices: tuple[int, ...] | tuple[tuple[int, ...], ...] | None = None
+
+    if site_list is not None:
+        # Extract raw site indices
+        raw_indices = [s.index() for s in site_list]
+
+        # For local multi-qubit operations, the flat list represents consecutive pairs
+        # due to reinterpret_cast from vector<pair<Site, Site>> to vector<Site>
+        if not is_zoned and qubits_num is not None and qubits_num > 1:
+            # Group consecutive elements into tuples of size qubits_num
+            if len(raw_indices) % qubits_num == 0:
+                site_tuples = [tuple(raw_indices[i : i + qubits_num]) for i in range(0, len(raw_indices), qubits_num)]
+                site_indices = tuple(site_tuples)
+            else:
+                # Fallback: treat as flat list if not evenly divisible
+                site_indices = tuple(sorted(set(raw_indices)))
+        else:
+            # For single-qubit and global operations, use flat list
+            site_indices = tuple(sorted(set(raw_indices)))
+
+    return DeviceOperationInfo(
+        name=name,
+        qubits_num=qubits_num,
+        parameters_num=op.parameters_num(),
+        duration=op.duration(),
+        fidelity=op.fidelity(),
+        interaction_radius=op.interaction_radius(),
+        blocking_radius=op.blocking_radius(),
+        idling_fidelity=op.idling_fidelity(),
+        is_zoned=is_zoned,
+        mean_shuttling_speed=op.mean_shuttling_speed(),
+        sites=site_indices,
+    )
+
+
+def _remap_operation_sites(
+    op_info: DeviceOperationInfo,
+    site_index_to_qubit_index: dict[int, int],
+) -> DeviceOperationInfo:
+    """Remap site indices in an operation to logical qubit indices.
+
+    Args:
+        op_info: Operation info with raw site indices.
+        site_index_to_qubit_index: Mapping from raw site indices to logical qubit indices.
+
+    Returns:
+        Operation info with remapped site indices.
+    """
```

Comment:
I am not 100% sure I get why this function is necessary. Could you explain?
The term "logical qubit indices" is also fairly dangerous and I'd rather avoid it. In no way are we talking about error correction here.

Reply:
Yes, this has sprung from the way NA device sites are indexed. @ystade raised the point somewhere in the discussions above. If I remember correctly, we agreed to work around it in this PR and fix it in the NA device later, as it does not belong to this PR.